src/org/sparkboard/server/server.clj

```diff
@@ -512,53 +320,58 @@
-                                              (case (:type payload)
-                                                "shortcut" (:callback_id payload "UNSET")
-                                                ("view_submission" "view_closed") (-> payload :view :callback_id)
-                                                "block_actions" (-> payload :actions first :action_id)))]
-                      [handler-id (assoc context :slack/payload payload)]))]
-          (log/info handler-id (select-keys context [:slack/user-id :slack/team-id ]))
-          (log/debug :context context)
-          (def LAST-CONTEXT context)
-          (.execute ^java.util.concurrent.ExecutorService pool
-                    ^Callable #(binding [slack/*request* context]
-                                 ((handlers handler-id (fn [& args] (log/error :unhandled-request handler-id args)))
-                                  (assoc context ::handler-id handler-id)))))
-        (ring.http/ok))))
+    (let [context
+          (cond event
+                (-> (slack-db/slack-context (:team_id params)
+                                            (if (map? (:user event)) ;; sometimes `user` is an id string, sometimes a map containing the id
+                                              (:id (:user event))
+                                              (:user event)))
+                    (assoc :slack/event event
+                           ::handler-id (keyword "event" (:type event))))
+
+                payload
+                (let [payload (-> (json->clj payload)
+                                  ;; private_metadata is deserialized here, and serialized in `slack.hiccup`
+                                  (update-in [:view :private_metadata] #(if (str/blank? %) nil (transit/read %)))
+                                  (set/rename-keys {:private_metadata :private_metadata}))]
+                  (-> (slack-db/slack-context (-> payload :team :id) (-> payload :user :id))
+                      (assoc :slack/payload payload
+                             ::handler-id (case (:type payload)
+                                            "shortcut" (keyword "shortcut" (:callback_id payload))
+                                            "view_submission" (keyword (-> payload :view :callback_id) "submit")
+                                            "view_closed" (keyword (-> payload :view :callback_id) "close")
+                                            "block_actions" (keyword (-> payload :actions first :action_id))
+                                            (keyword "UNKNOWN_PAYLOAD_TYPE" (:type payload "nil")))))))
+          handler (handlers (::handler-id context) missing-handler)
+          exec #(binding [slack/*context* context] (handler context))]
+      (log/debug (::handler-id context) context)
+      (if (:response-action? (meta handler))
+        ;; eval & return result
+        (or (exec)
+            (ring.http/ok))
```

Comment:
How would we reach the second possibility of this `or`? Is the rule that handlers return an HTTP response or `nil`?

Reply:
yeah, that was the idea. alternatively one could require the handler to return `http/ok`